cubeb-core/src/ffi_types.rs

```diff
@@ -0,0 +1,219 @@
+                    unsafe { $drop(self.0) }
+                }
+            }
+        )*
+
+        $(
+            impl Clone for $owned {
+                #[inline]
+                fn clone(&self) -> $owned {
+                    unsafe {
+                        let handle: *mut $ctype = $clone(self.0);
+                        ::foreign_types::ForeignType::from_ptr(handle)
+                    }
+                }
+            }
+
+            impl ::std::borrow::ToOwned for $borrowed {
+                type Owned = $owned;
+                #[inline]
+                fn to_owned(&self) -> $owned {
+                    unsafe {
+                        let handle: *mut $ctype = $clone(::foreign_types::ForeignTypeRef::as_ptr(self));
+                        ::foreign_types::ForeignType::from_ptr(handle)
+                    }
+                }
+            }
+        )*
+
+        impl ::std::ops::Deref for $owned {
+            type Target = $borrowed;
+
+            #[inline]
+            fn deref(&self) -> &$borrowed {
+                unsafe { ::foreign_types::ForeignTypeRef::from_ptr(self.0) }
+            }
+        }
+
+        impl ::std::ops::DerefMut for $owned {
+            #[inline]
+            fn deref_mut(&mut self) -> &mut $borrowed {
+                unsafe { ::foreign_types::ForeignTypeRef::from_ptr_mut(self.0) }
+            }
+        }
+
+        impl ::std::borrow::Borrow<$borrowed> for $owned {
+            #[inline]
+            fn borrow(&self) -> &$borrowed {
+                &**self
+            }
+        }
+
+        impl ::std::convert::AsRef<$borrowed> for $owned {
+            #[inline]
+            fn as_ref(&self) -> &$borrowed {
+                &**self
+            }
+        }
+
+        $(#[$borrowed_attr])*
+        pub struct $borrowed(::foreign_types::Opaque);
+
+        $(#[$impl_attr])*
+        impl ::foreign_types::ForeignTypeRef for $borrowed {
+            type CType = $ctype;
+        }
+
+        impl ::std::fmt::Debug for $borrowed {
+            fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+                let ptr = self as *const $borrowed as usize;
+                f.debug_tuple(stringify!($borrowed))
+                    .field(&ptr)
+                    .finish()
+            }
+        }
+    }
+}
+
+/// A macro to define a wrapper around a POD FFI type that lives on
+/// the stack.
+macro_rules! ffi_type_stack {
+    ($(#[$impl_attr:meta])*
+     type CType = $ctype:ty;
+     $(#[$owned_attr:meta])*
+     pub struct $owned:ident;
+     $(#[$borrowed_attr:meta])*
+     pub struct $borrowed:ident;
+    ) => {
+        $(#[$owned_attr])*
+        pub struct $owned($ctype);
+
+        $(#[$impl_attr])*
+        impl ::foreign_types::ForeignType for $owned {
+            type CType = $ctype;
+            type Ref = $borrowed;
+
+            unsafe fn from_ptr(_: *mut $ctype) -> $owned {
+                // $owned(*ptr)
+                panic!("Not implemented.")
```

Comment:
Is this intended to be left unimplemented?

Reply:
Yes. Maybe I should use the special macro [std::unimplemented!](https://doc.rust-lang.org/std/macro.unimplemented.html)?

I'm not completely happy with the `foreign_types` crate. Have experiments with just implementing that macros to generate all the boilerplate code without the need for `ForeignType` and `ForeignTypeRef` traits. Having to re-export those two traits and `use` them in the code is a minor nit. But removing the reliance on `foreign_types`, the user can just `use cubeb::ContextRef`, for example.

Reply:
> Yes. Maybe I should use the special macro std::unimplemented!?

I think that'd be clearer (and removing the commented out code above it).

> Having to re-export those two traits and use them in the code is a minor nit.

Yeah, I was a bit surprised when I saw that initially... it makes sense, but it did feel a bit suboptimal.

Reply:
OK, I'll take this opportunity to replace `foreign_types` with the macros I wrote. This will clean up the nits and won't need the `unimplemented!` function to satisfy the trait.